### PR TITLE
Fix sub region ids

### LINKED ISSUE #6: South Australian subregion IDs are wrong in sub_regional_reference_nodes.csv
In sub_regional_reference_nodes.csv there are the lines:

South Australia,Central South Australia (SA),Torrens Island 66 kV
South Australia,South East South Australia (CESA),South East 132 kV

Indicating sub region IDs of SA in CESA, but in other tables in the work book and in the trace data the IDs CSA and SESA are used.

### LINKED ISSUE #7: Mapping to sub_regions when regional granularity is specified
At the moment with the regional template, we map to just a single subregion, which makes sense. However, when building the demand traces, I think we'd need to sum across all of the subregional nodes in a region to get the total for the region, which begs the question if the node input file doesn't have this mapping where does the translator get it from? Maybe the templater needs to also spit out a separate mapping file that is always created specifying the full mapping between region, subregion, and rez?

### PR DESCRIPTION
**Pull request by nick-gorman**
- implements fuzzy match for sub regions names to a set of hardcoded names and then sets sub regions IDs based on hardcoded mapping from names to IDs. This is the same approach as used for nem regions (fixes #6)
- Add a region_and_zone mapping table to the template, which will allow for mapping of nem regions, sub regions, and rezs independently of the node table (fixes #7)

### CODE REVIEW

**Comment by prakaa** [comment]
Closes #6, #7

**Comment by prakaa** [comment]
>     * Add a region_and_zone mapping table to the template, which will allow for mapping of nem regions, sub regions, and rezs independently of the node table (fixes [Mapping to sub_regions when regional granularity is specified #7](https://github.com/Open-ISP/ISPyPSA/issues/7))

The REZ mapping table doesn't include a small sub-region:  GG (Gladstone Grid). Could you still load the REZ table and then add sub-regions generated by `_template_sub_regional_nodes` that aren't in the REZ table?

**Comment by nick-gorman** [comment]
> > ```
> > * Add a region_and_zone mapping table to the template, which will allow for mapping of nem regions, sub regions, and rezs independently of the node table (fixes [Mapping to sub_regions when regional granularity is specified #7](https://github.com/Open-ISP/ISPyPSA/issues/7))
> > ```
> 
> The REZ mapping table doesn't include a small sub-region: GG (Gladstone Grid). Could you still load the REZ table and then add sub-regions generated by `_template_sub_regional_nodes` that aren't in the REZ table?

What about keeping the rez table as just a rez table, rather than introducing sub-regions with no rezs, and going with your original suggestion of always templating the nodes at a sub_regional granularity so that can be used from mapping from region to sub region?

File: src/ispypsa/templater/nodes.py [hunk 1]
```diff
@@ -147,40 +126,60 @@ def _template_regional_node_table(
...
         ],
         axis=1,
     )
-    regional_nodes["nem_region"] = _fuzzy_match_names(
-        regional_nodes["nem_region"],
+    regional_nodes = _match_region_name_and_id(regional_nodes)
+    return regional_nodes
+
+
+def _split_out_sub_region_name_and_id(data):
```
Comment (prakaa) [inline src/ispypsa/templater/nodes.py:144]: Declare types in function definition

File: src/ispypsa/templater/nodes.py [hunk 2]
```diff
@@ -147,40 +126,60 @@ def _template_regional_node_table(
...
+    sub_region_name_and_id.columns = [_snakecase_string(name_id_col)]
+    sub_region_name_and_id[_snakecase_string(name_id_col + " ID")] = (
+        sub_region_name_and_id[
+            _snakecase_string(name_id_col)
+        ].replace(_NEM_SUB_REGION_IDS)
+    )
+    return sub_region_name_and_id
+
+
+def _match_region_name_and_id(data):
```
Comment (prakaa) [inline src/ispypsa/templater/nodes.py:163]: Declare types

File: src/ispypsa/templater/nodes.py [hunk 3]
```diff
@@ -147,40 +126,60 @@ def _template_regional_node_table(
...
         _NEM_REGION_IDS.keys(),
         "determining the NEM region",
     )
-    regional_nodes["nem_region_id"] = regional_nodes["nem_region"].replace(
-        _NEM_REGION_IDS
+    data["nem_region_id"] = data["nem_region"].replace(_NEM_REGION_IDS)
+    return data
+
+
+def _extract_voltage(data, column):
```
Comment (prakaa) [inline src/ispypsa/templater/nodes.py:173]: Declare types

**Comment by prakaa** [comment]
> > > ```
> > > * Add a region_and_zone mapping table to the template, which will allow for mapping of nem regions, sub regions, and rezs independently of the node table (fixes [Mapping to sub_regions when regional granularity is specified #7](https://github.com/Open-ISP/ISPyPSA/issues/7))
> > > ```
> > 
> > 
> > The REZ mapping table doesn't include a small sub-region: GG (Gladstone Grid). Could you still load the REZ table and then add sub-regions generated by `_template_sub_regional_nodes` that aren't in the REZ table?
> 
> What about keeping the rez table as just a rez table, rather than introducing sub-regions with no rezs, and going with your original suggestion of always templating the nodes at a sub_regional granularity so that can be used from mapping from region to sub region?

Makes sense, but template should reflect the actual granularity. There should be a separate mapping if the granularity is not sub_regional and this effectively the sub_regional node table.

So to be clear:

1. `granularity=sub_regional`: Produces a `nodes.csv`
2. `granularity=regional`: Produces a `nodes.csv` with regional nodes, and an additional `subregion_region_mapping.csv` or something similar that is very similar to a sub_regional nodes.csv, but perhaps with cols dropped

File: src/ispypsa/templater/nodes.py [hunk 4]
```diff
@@ -147,40 +126,60 @@ def _template_regional_node_table(
...
         axis=1,
     )
-    regional_nodes["nem_region"] = _fuzzy_match_names(
-        regional_nodes["nem_region"],
+    regional_nodes = _match_region_name_and_id(regional_nodes)
+    return regional_nodes
+
+
+def _split_out_sub_region_name_and_id(data):
+    print(data)
```
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:145]: agreed

File: src/ispypsa/templater/regions_and_zones.py
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/regions_and_zones.py:1]: going to keep it here and just always template sub regional nodes for the purpose of mapping sub regional demand data to the correct nem region.

File: src/ispypsa/templater/mappings.py [hunk 5]
```diff
@@ -17,6 +17,24 @@
...
 
+_NEM_SUB_REGION_IDS = pd.Series(
+    {
+        "Northern Queensland": "NQ",
+        "Central Queensland": "CQ",
+        "Gladstone Grid": "GG",
+        "South Queensland": "SQ",
+        "Northern New South Wales": "NNSW",
+        "Central New South Wales": "CNSW",
+        "South NSW": "SNSW",
```
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/mappings.py:28]: argeed

File: src/ispypsa/templater/nodes.py [hunk 6]
```diff
@@ -252,6 +251,17 @@ def _split_name_id(series: pd.Series) -> pd.DataFrame:
     return split_name_id
 
 
+def _capture_just_name(series: pd.Series) -> pd.DataFrame:
+    """
+    Capture the name (plain English) and not the ID in parentheses (capitalised letters)
+    using a regular expression on a string `pandas.Series`.
+    """
+    split_name_id = series.str.strip().str.extract(
+        r"(?P<name>[A-Za-z\s,]+)(?=\s\([A-Z]+\))", expand=True
```
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:260]: agreed

File: src/ispypsa/templater/nodes.py (hunk 1 above)
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:144]: agreed

File: src/ispypsa/templater/nodes.py (hunk 3 above)
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:173]: agreed

**Comment by nick-gorman** [comment]
> > > > ```
> > > > * Add a region_and_zone mapping table to the template, which will allow for mapping of nem regions, sub regions, and rezs independently of the node table (fixes [Mapping to sub_regions when regional granularity is specified #7](https://github.com/Open-ISP/ISPyPSA/issues/7))
> > > > ```
> > > 
> > > 
> > > The REZ mapping table doesn't include a small sub-region: GG (Gladstone Grid). Could you still load the REZ table and then add sub-regions generated by `_template_sub_regional_nodes` that aren't in the REZ table?
> > 
> > 
> > What about keeping the rez table as just a rez table, rather than introducing sub-regions with no rezs, and going with your original suggestion of always templating the nodes at a sub_regional granularity so that can be used from mapping from region to sub region?
> 
> Makes sense, but template should reflect the actual granularity. There should be a separate mapping if the granularity is not sub_regional and this effectively the sub_regional node table.
> 
> So to be clear:
> 
> 1. `granularity=sub_regional`: Produces a `nodes.csv`
> 2. `granularity=regional`: Produces a `nodes.csv` with regional nodes, and an additional `subregion_region_mapping.csv` or something similar that is very similar to a sub_regional nodes.csv, but perhaps with cols dropped

Thats correct.

File: src/ispypsa/templater/mappings.py [hunk 7]
```diff
@@ -17,6 +17,24 @@
...
+        "Sydney, New Castle, Wollongong": "SNW",
```
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/mappings.py:29]: agreed

File: src/ispypsa/templater/nodes.py [hunk 8]
```diff
@@ -147,40 +126,60 @@ def _template_regional_node_table(
...
-        regional_nodes["nem_region"],
+    regional_nodes = _match_region_name_and_id(regional_nodes)
+    return regional_nodes
+
+
+def _split_out_sub_region_name_and_id(data):
+    print(data)
+    name_id_col = "ISP Sub-region"
+    sub_region_name_and_id = _capture_just_name(data[name_id_col])
+    print(sub_region_name_and_id["name"])
```
[the remark this answers is not in the document]
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:148]: agreed

File: src/ispypsa/templater/nodes.py (hunk 2 above)
Comment (nick-gorman) [reply src/ispypsa/templater/nodes.py:163]: agreed

**Comment by nick-gorman** [comment]
Somethings gone wrong with dodo.py. Let me take a look.

**Pull request merged**